### ISSUE TITLE
Enable BatchNorm to use running mean/variance during train

### ISSUE THREAD

**Issue by GuillaumeTong** [pytorch/pytorch#64730]
## 🚀 Feature

As the title says:
Enable BatchNorm to use running mean/variance during train

## Motivation

There are cases where the mean/variance of one piece of data going through a BatchNorm layer is not representative of the mean/variance of all data it expects to receive, such as the following cases:

1. **Small batch size**: mean/variance calculated from 2 samples will almost always be skewed, yet a small batch size could be necessary depending on the model
2. **Different types of data all go through the same layer instance**: When training some types of model, it becomes necessary to feed different sets of input through the model one-by-one (such as, for example, when using a GAN cycle-type loss where the data stream goes: noise --decoder--> image --encoder--> embedding --decoder--> image2). In these kinds of cases, some BatchNorm layers of such models experience two different distributions with different mean/variance, but will normalize them into the same distribution during train (since the stats of each distribution is taken), while normalizing them differently during eval (due to using the estimated stats of the union of both distribution).

## Pitch

In order of importance as I see it currently, this is what is needed:
1.  Enable BatchNorm to use some form of running mean/variance during train, with an optional argument that can default to preserve current behavior
2.  The stats could be calculated from a sliding window, so that different sets of data can have equal weight (for the case where different sets of data have to go through the same layer within the same batch)
3. Create a convenience function to apply this setting to all batchnorm layers of a network, so that pretrained models built without this consideration can be retrofitted.

## Alternatives

Do not use BatchNorm?

## Additional context

I have recreated a minimal case where this issue can be observed in this notebook (I might not keep it public forever on my own google account, please advise on how best to permanently share a jupyter notebook):
https://colab.research.google.com/drive/16heLOdeL61e2_X7BD9Xotv05BG4QKVGF?usp=sharing

You will be able to see that in my extreme setup, the network should theoretically only be able to reach 75% accuracy, but actually performs 100% train accuracy with only 25% val accuracy

**Comment by GuillaumeTong** [comment]
Any feedback? Does it make sense to ask for this here?

**Comment by GuillaumeTong** [comment]
It is similar, yes, but not exactly the same I think:
- #66073 wants to use its feature to do a separate finetuning on specific data, while I would like here to start from the very beginning with the "aggregated mode" (tentative name?).
- #66073 wants to keep statistics in BN constant from the more general dataset, but I'd like to keep updating the estimated stats as we iterate through the dataset, in order to leverage the main advantage of BN: use the normalization to combat internal covariate shift (ICS).

Overall, I am advocating for a hybrid behavior hybrid between train and val: use the estimated stats for normalizing, just like in val, but also update the stats, like in train.
I also think that in this case, in order to make the statistics during train more agile and able to react to IDS, it could be that the normalization during train should act based on statistics from a window of batches to **simulate a bigger batch size**, instead of exponentially averaged statistics used in val.

**Comment by albanD** [comment]
Thanks for the details!

I think there are two things:
- The track_running_stats and training parameters are quite confusing today. But they would allow you to do what you want. The main issue here is that the training param is linked to the Module's state. And maintaining that to eval within a network that is in training mode is tricky.
- Larger improvements on how statistics are tracked (based on window for example).

Is that correct?
